src/app/clusters/operational-state-server/operational-state-delegate.h

```diff
@@ -0,0 +1,393 @@
+        if (ph.isNullable())
+        {
+            set();
+        }
+        else
+        {
+            set(ph.PhaseName, sizeof(ph.PhaseName));
+        }
+        return *this;
+    }
+
+    void set(const char * name = nullptr, size_t nameLen = 0)
+    {
+        if (name == nullptr)
+        {
+            phaseName.SetNull();
+        }
+        else
+        {
+            memset(PhaseName, 0, sizeof(PhaseName));
+            if (nameLen > kOperationalPhaseNameMaxSize)
+            {
+                memcpy(PhaseName, name, kOperationalPhaseNameMaxSize);
+            }
+            else
+            {
+                memcpy(PhaseName, name, nameLen);
+            }
+            phaseName = chip::app::DataModel::Nullable<chip::CharSpan>(chip::CharSpan(PhaseName, sizeof(PhaseName)));
+        }
+    }
+    bool isNullable() const { return phaseName.IsNull(); }
+};
+
+/**
+ * A class hold the phase list of operational state cluster
+ */
+struct GenericOperationalPhaseList : public GenericOperationalPhase
+{
+    GenericOperationalPhaseList(const char * name = nullptr, size_t nameLen = 0) : GenericOperationalPhase(name, nameLen) {}
+    GenericOperationalPhaseList * next = nullptr;
+};
+
+/**
+ * A class wrap the operation completion of operational state cluster
+ */
+struct GenericOperationCompletion : public chip::app::Clusters::OperationalState::Events::OperationCompletion::Type
+{
+    GenericOperationCompletion(uint8_t aCompletionErrorCode)
+    {
+
+        chip::app::DataModel::Nullable<uint32_t> __totalOperationalTime;
+        chip::app::DataModel::Nullable<uint32_t> __pausedTime;
+
+        completionErrorCode = aCompletionErrorCode;
+
+        __totalOperationalTime.SetNull();
+        totalOperationalTime.SetValue(__totalOperationalTime);
+
+        __pausedTime.SetNull();
+        pausedTime.SetValue(__pausedTime);
+    }
+
+    GenericOperationCompletion(uint8_t aCompletionErrorCode, uint32_t aTotalOperationalTime)
+    {
+        chip::app::DataModel::Nullable<uint32_t> __pausedTime;
+
+        completionErrorCode = aCompletionErrorCode;
+        totalOperationalTime.SetValue(chip::app::DataModel::Nullable<uint32_t>(aTotalOperationalTime));
+        __pausedTime.SetNull();
+        pausedTime.SetValue(__pausedTime);
+    }
+
+    GenericOperationCompletion(uint8_t aCompletionErrorCode, uint32_t aTotalOperationalTime, uint32_t aPausedTime)
+    {
+        completionErrorCode = aCompletionErrorCode;
+        totalOperationalTime.SetValue(chip::app::DataModel::Nullable<uint32_t>(aTotalOperationalTime));
+        pausedTime.SetValue(chip::app::DataModel::Nullable<uint32_t>(aPausedTime));
+    }
+};
+
+/**
+ * A delegate to handle application logic of the Operational State aliased Cluster.
+ */
+class Delegate
+{
+public:
+    /**
+     * Get operational state.
+     * @param void.
+     * @return the const reference of operational state.
+     */
+    virtual const GenericOperationalState & GetOperationalState() const = 0;
```

Comment:
This requies the delegate to keep the state in RAM, right?  That seems a little unfortunate.